ethcore/private-tx/src/log.rs

```diff
@@ -0,0 +1,247 @@
+			validators: validator_logs,
+			deployment_timestamp: None,
+			public_tx_hash: None,
+		});
+	}
+
+	/// Logs the obtaining of the signature for the private transaction
+	pub fn signature_added(&self, tx_hash: H256, validator: Address) {
+		let mut logs = self.logs.write();
+		if let Some(transaction_log) = logs.get_mut(&tx_hash) {
+			transaction_log.status = PrivateTxStatus::Validating;
+			if let Some(ref mut validator_log) = transaction_log.validators.iter_mut().find(|log| log.account == validator) {
+				validator_log.validated = true;
+				validator_log.validation_timestamp = Some(SystemTime::now().duration_since(UNIX_EPOCH).unwrap_or_default().as_secs());
+			}
+		}
+	}
+
+	/// Logs the final deployment of the resulting public transaction
+	pub fn tx_deployed(&self, tx_hash: H256, public_tx_hash: H256) {
+		let mut logs = self.logs.write();
+		if let Some(log) = logs.get_mut(&tx_hash) {
+			log.status = PrivateTxStatus::Deployed;
+			log.deployment_timestamp = Some(SystemTime::now().duration_since(UNIX_EPOCH).unwrap_or_default().as_secs());
+			log.public_tx_hash = Some(public_tx_hash);
+		}
+	}
+
+	fn read_logs(&self) {
+		let log_file = match self.logs_dir {
+			Some(ref path) => {
+				let mut file_path = path.clone();
+				file_path.push("private_tx.log");
+				match File::open(&file_path) {
+					Ok(file) => file,
+					Err(err) => {
+						trace!(target: "privatetx", "Cannot open logs file: {}", err);
+						return;
+					}
+				}
+			}
+			None => {
+				warn!(target: "privatetx", "Logs path is not defined");
+				return;
+			}
+		};
+		let mut transaction_logs: Vec<TransactionLog> = match serde_json::from_reader(log_file) {
+			Ok(logs) => logs,
+			Err(err) => {
+				error!(target: "privatetx", "Cannot deserialize logs from file: {}", err);
+				return;
+			}
+		};
+		// Drop old logs
+		let current_timestamp = SystemTime::now().duration_since(UNIX_EPOCH).unwrap_or_default().as_secs();
+		transaction_logs.retain(|tx_log| current_timestamp - tx_log.creation_timestamp < MAX_STORING_TIME);
+		let mut logs = self.logs.write();
+		for log in transaction_logs {
+			logs.insert(log.tx_hash, log);
+		}
+	}
+
+	fn flush_logs(&self) {
+		if self.logs.read().is_empty() {
+			// Do not create empty file
+			return;
+		}
+		let log_file = match self.logs_dir {
+			Some(ref path) => {
+				let mut file_path = path.clone();
+				file_path.push("private_tx.log");
+				match File::create(&file_path) {
+					Ok(file) => file,
+					Err(err) => {
+						trace!(target: "privatetx", "Cannot open logs file for writing: {}", err);
+						return;
+					}
+				}
+			}
+			None => {
+				return;
+			}
+		};
+		if let Err(err) = serde_json::to_writer(log_file, &self.logs.read().values().cloned().collect::<Vec<TransactionLog>>()) {
+			error!(target: "privatetx", "Error during logs serialisation: {}", err);
+		}
+	}
+}
+
+// Flush all logs on drop
+impl Drop for Logging {
+	fn drop(&mut self) {
+		self.flush_logs();
```

Comment:
I am not sure if writing the JSON file in a destructor is the best approach. One can not return errors back and as a destructor it executes quite slowly. Would it not be possible to offer explicit read/write functionality in the logging API?

Reply:
I was heavily thinking between these two approaches (log every event into the file vs flush all logs in the destructor). The problem with logging every event is, that these events are not sequential, as a result I would need to rewrite the whole log file in order to update one log record. Also not very efficient. So I've decided to stop on the flush in the dtor. But I'm open to the discussion ;-) and I've also refactored this part and added a test for serialization in order to proceed with any changes easily.

Reply:
It was only concerned with offering explicit read/write functions which return `Result` instead of reading and writing in constructor and destructor with no way to report errors. The granularity of writes should be unaffected.

Reply:
Discussed this moment offline. As a result, I would like to re-phrase @twittner 's concern (as I understand its application): should we return the error during RPC call of private transaction methods in the case, when flushing\reading logs during this call fails. My opinion is that we should not: 1) logging is just a secondary functionality and the primary functionality must not suffer from its flaws. The informing of the client via error warnings is enough IMO for such cases, 2) half of the cases, when private transaction methods are called, is the automatic processing of incoming packets. In this case returning the error doesn't make any sense at all, because it cannot be handled by anybody.